### PR TITLE
feat!: use `distinct` return value witnesses by default

### PR DESCRIPTION
**Pull request by TomAFrench**
# Description

## Problem\*

Resolves https://github.com/noir-lang/noir/issues/4914
Resolves https://github.com/noir-lang/noir/issues/4443

## Summary\*

We're starting to need various workarounds for non-distinct witnesses to the point where it's more of a hindrance than it's worth. This PR makes the `distinct` behaviour the default and raises a parser error when using the `distinct` keyword.

I've gone for a hard error rather than a warning as the only place where `distinct` can be used is on `main` and so there's no risk of breaking libraries. Affected users can simply modify their own binary project to fix the issue.

## Documentation\*

Check one:
- [ ] No documentation needed.
- [x] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: compiler/noirc_frontend/src/parser/parser/function.rs [hunk 1]
```diff
@@ -174,7 +184,7 @@ mod test {
                 "fn f(f: pub Field, y : Field, z : Field) -> u8 { x + a }",
                 "fn f<T>(f: pub Field, y : T, z : Field) -> u8 { x + a }",
                 "fn func_name(x: [Field], y : [Field;2],y : pub [Field;2], z : pub [u8;5])  {}",
-                "fn main(x: pub u8, y: pub u8) -> distinct pub [u8; 2] { [x, y] }",
```
Comment (vezenovm) [inline compiler/noirc_frontend/src/parser/parser/function.rs:177]: Could add this into the failing definitions to parse below

File: tooling/nargo_fmt/tests/input/fn.nr [hunk 2]
```diff
@@ -19,28 +19,45 @@ fn main(
 
 fn main<T>() where T: Eq {}
 
-fn main(tape: [Field; TAPE_LEN], initial_registers: [Field; REGISTER_COUNT], initial_memory: [Field; MEM_COUNT], initial_program_counter: Field, initial_call_stack: [Field; MAX_CALL_STACK], initial_call_stack_pointer: u64) -> pub ExecutionResult {}
-
-fn apply_binary_field_op<N>(lhs: RegisterIndex, rhs: RegisterIndex, result: RegisterIndex, op: u8, registers: &mut Registers<N>) -> bool {}
-
-fn main() -> distinct pub [Field;2] {}
+fn main(
```
Comment (vezenovm) [inline tooling/nargo_fmt/tests/input/fn.nr:22]: Why did this change? Is it not input?

File: compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs [hunk 3]
```diff
@@ -335,28 +333,36 @@ impl Ssa {
...
+    // This can be tackled by:
+    // - Tracking the last assigned public input witness and only renumbering a witness if it is below this value.
+    // - Modifying existing constraints to rearrange their outputs so they are suitable
+    //   - See: https://github.com/noir-lang/noir/pull/4467
+    let distinct_return_witness: Vec<_> = acir
+        .return_witnesses
+        .clone()
+        .into_iter()
+        .map(|return_witness| acir.create_witness_for_expression(&Expression::from(return_witness)))
+        .collect();
```
Comment (jfecher) [inline compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs:361]:
```suggestion
    let distinct_return_witness = vecmap(acir.return_witnesses.clone(), |return_witness| {
        acir.create_witness_for_expression(&Expression::from(return_witness))
    });
```

**Review by vezenovm — APPROVED** [review]
Approving, but I was surprised by the fmt change if we could address that in a follow up

File: tooling/nargo_fmt/tests/input/fn.nr (hunk 2 above)
Comment (TomAFrench) [reply tooling/nargo_fmt/tests/input/fn.nr]: As `distinct` causes an error having it here prevents the test from running properly.

Comment (TomAFrench) [reply tooling/nargo_fmt/tests/input/fn.nr]: I've reverted the other changes caused by automatic formatting when saving the file.

File: cspell.json [hunk 4]
```diff
@@ -54,6 +54,7 @@
         "csat",
         "curvegroup",
         "databus",
+        "datastructures",
```
Comment (jfecher) [inline cspell.json:57]: Shouldn't this be two words?

**Review by jfecher — APPROVED** [review]
LGTM

**Review by vezenovm — APPROVED** [review]

**Pull request merged**